### ISSUE TITLE
java.lang.AbstractMethodError after creating class inheriting from trait

### ISSUE THREAD

**Issue by Alex-At-Home** [lihaoyi/Ammonite#476]
Mentioned this in [gitter](https://gitter.im/lihaoyi/Ammonite) also, looking for some quick suggestions on how to investigate further.

Also apologies since I didn't (yet) find an easy way to reproduce this with a small fragment of code.

Summary:
- I have a completely boring trait ([here](https://github.com/Alex-At-Home/elasticsearch_scala_driver/blob/6c75a2c3db7ba8ce0aac3db8f0f4b147d8c0a76e/rest_scala_core/src/main/scala/org/elastic/rest/scala/driver/RestBase.scala#L230)), with some abstract members and some concrete ones
- If I create a sub-class that fills in the abstract members (see below for example), then if I try to access any member (originally abstract or not), I "usually" get an error like `java.lang.AbstractMethodError: $sess.cmd99$Internal5.getUrl()Ljava/lang/String;`
  - _(worth noting that the same code works fine when run normally in Scala, ie it's tested working logic)_
- Exceptions to "usually":
  - If the very first operation is creating an instance of the sub-class, then it always works
  - I have tried copy-pasting the trait into Ammonite, and it always works
  - I created an identical copy in my codebase (renamed `BaseDriverOpX`) and it works (of course the new trait does not have dependencies, so presumably the compile order etc is different)
- (I am starting Ammonite in what I believe to be [a super standard way](https://github.com/Alex-At-Home/elasticsearch_scala_driver/blob/2817f944313008f139993e9b028b9befa60a1b6c/elasticsearch_scala_shell/src/main/scala/org/elastic/elasticsearch/scala/driver/jvm/ShellMain.scala))

Example of the sub-class I create:

``` scala
case class Internal
   (resource: RestResource, op: String, body: Option[String], mods: List[(String, Any)], headers: List[String]) 
   extends BaseDriverOp
{
   override def withModifier(kv: (String, Any)): this.type = Internal(resource, op, body, kv :: mods, headers)
            .asInstanceOf[this.type]

   override def withHeader(h: String): this.type = Internal(resource, op, body, mods, h :: headers)
            .asInstanceOf[this.type]
}
```

And what happens:

``` scala
@ Internal(latest.`/`(), "GET", None, List(), List()).getUrl 
java.lang.AbstractMethodError: $sess.cmd99$Internal.getUrl()Ljava/lang/String;
  $sess.cmd102$.<init>(cmd102.sc:1)
  $sess.cmd102$.<clinit>(cmd102.sc:-1)
```

Unless I define the class as the very first call in the new shell, or I sub-class from an identical copy of the trait, in which case:

``` scala
@ Internal(latest.`/`(), "GET", None, List(), List()) 
res97: Internal = Internal(/(), "GET", None, List(), List())
@ res97.getUrl 
res98: String = "/"
```

(it doesn't matter whether I instantiate the object and then call `getUrl` in 2 steps or 1)

Anyone have any thoughts or ideas for areas/steps to investigate futher? Thanks in advance

**Comment by Alex-At-Home** [comment]
A bit more investigation:
* https://issues.scala-lang.org/browse/SI-9689 seems related
* I added `repl.compiler.settings.YnoLoadImplClass.tryToSetFromPropertyValue("true")` to the predef, and confirmed it took:

```scala
@ repl.compiler.settings.YnoLoadImplClass  
res0: tools.nsc.Settings#BooleanSetting = -Yno-load-impl-class = true
```

Unfortunately that does not solve the problem

**Comment by Alex-At-Home** [comment]
@lihaoyi is there a way of examining the `$sess.cmd` objects? Eg viewing their scala or AST or bytecode?

I need to figure out why  `$sess.cmd0.Internal(latest.`/`(), "GET", None, List(), List())` `$sess.cmd4.Internal(latest.`/`(), "GET", None, List(), List())` behave differently despite being the identical class pasted into the REPL twice...

**Comment by lihaoyi** [comment]
@Alex-At-Home there isn't a way of looking at the source of `$sess.cmd` in the published version, but if you want to check out the repo and build/run it from source you could add a `println` over here:

- https://github.com/lihaoyi/Ammonite/blob/master/amm/interp/src/main/scala/ammonite/runtime/Interpreter.scala#L258

Which would tell you *exactly* what Scala source-code-strings are being assembled, compiled and executed

**Comment by Alex-At-Home** [comment]
OK I've made a bit more progress with this.

I can reproduce _and fix_ it with sbt console, but the same fix doesn't seem to work with Ammonite, I think I just need to inject a setting _somewhere_ but I'm not sure where.

Here are the most recent findings:
* If from sbt I run `elasticsearch_scala_shell/console`, I get exactly the same (above) behavior as with Ammonite, eg the first call to (eg) `latest().read().getUrl` returns the expected string and the second call throws `AbstractMethodError`
* BUT ... If I add `scalacOptions ++= Seq("-Yno-load-impl-class")` to `elasticsearch_scala_shell`s settings (a basically empty project that `dependsOn` all the code in) then it all starts working fine, ie all calls to (eg) `latest().read().getUrl` the expected strings ... so good, and confirmed that the problem _is_ at least heavily related to https://issues.scala-lang.org/browse/SI-9689 
* However (unsurprisingly) the Ammonite JAR generated from `sbt  elasticsearch_scala_shell/assembly` still fails
* I say unsurprising because presumably `sbt console` is inheriting its project's `scalacOptions` when building code but Ammonite knows nothing about that
* So the question is why the [origin fix (a few comments above)](https://github.com/lihaoyi/Ammonite/issues/476#issuecomment-250569623) _didn't_ work

So @lihaoyi have you confirmed that you use `repl.compiler.settings.*` (but in particular `YnoLoadImplClass`!) in the code you compile? Can you think of any other differences between pasting the code into `sbt <project>/console` (with `scalacOptions` set in `Build.scala`) and into `java -jar <project>.jar MyAmmoniteMain` with `repl.compiler.settings.<custom settings>` in `predef.sc`?

**Comment by lihaoyi** [comment]
Yes, we use `repl.compiler.settings.<custom settings>` in the code we compile inside Ammonite. This is tested and relied upon by many people. Unless `YnoLoadImplClass` is special in some way and can't be initialized the same way as others

Ammonite currently does not support SBT `scalacOptions`. You need to set them using `repl.compiler.setting`, or via the soon-to-be introduced `interp.configureCompiler` API https://github.com/lihaoyi/Ammonite/commit/c6a282d4209130899dbea41aeab5a5ae46d1dcd7#diff-269c8021a73487ad843d4df1298e0a81R45

I'm really not sure what your issue is, but this really needs a minimized example; otherwise you're really just talking to yourself and possibly debugging in circles and I can't really contribute much at all. 

Feel free to ping this issue or open a new one when you've minimized it further 😄

**Comment by Alex-At-Home** [comment]
@lihaoyi ha I hadn't had time to play with this. but your brutal closing of my issue ( :stuck_out_tongue:  ) spurred me to action and I created a minimal example (which also highlighted what the issue was)

https://github.com/Alex-At-Home/ammonite-failure-case

Just clone and run `sbt clean test`, consistently fails for me

All (?) I'm doing it creating concrete classes from 2 traits defined [here](https://github.com/Alex-At-Home/ammonite-failure-case/blob/master/src/main/scala/example/Traits.scala), and then trying to access a method in the 2nd.

The [tests](https://github.com/Alex-At-Home/ammonite-failure-case/blob/master/src/test/scala/example/AmmoniteFailureSpec.scala) just performs the above operations twice, once in the code (which works) and once in the shell (which fails)

Any thoughts?

**Comment by lihaoyi** [comment]
Here's the minimized version of what you've given me:

project/build.sbt
```scala
sbt.version=0.13.13
```

build.sbt
```scala
scalaVersion := "2.11.8"

libraryDependencies += "com.lihaoyi" % "ammonite" % "0.9.0" % "test" cross CrossVersion.full

sourceGenerators in Test += Def.task {
  val file = (sourceManaged in Test).value / "amm.scala"
  IO.write(file, """object amm extends App { ammonite.Main().run() }""")
  Seq(file)
}.taskValue
```

Traits.scala
```scala
package example

object Traits {
    trait MiscTrait {
        val anotherParam: String
    } 
    trait BaseTrait {
        val param0: MiscTrait
        def showAnotherParam: String = param0.anotherParam 
    }
}
```

Console session:

```scala
lihaoyi ammonite-failure-case$ sbt test:run
[info] Loading project definition from /Users/lihaoyi/ammonite-failure-case/project
[info] Set current project to ammonite-failure-case (in build file:/Users/lihaoyi/ammonite-failure-case/)
[info] Updating {file:/Users/lihaoyi/ammonite-failure-case/}ammonite-failure-case...
[info] Resolving jline#jline;2.12.1 ...
[info] Done updating.
[info] Running amm
Welcome to the Ammonite Repl 0.9.0
(Scala 2.11.8 Java 1.8.0_112)
If you like Ammonite, please support our development at www.patreon.com/lihaoyi
@ case class Dummy(anotherParam: String) extends example.Traits.MiscTrait
defined class Dummy
@ case class Internal(param0: example.Traits.MiscTrait) extends example.Traits.BaseTrait
defined class Internal
@ Internal(Dummy("dummy")).showAnotherParam
java.lang.AbstractMethodError: $sess.cmd1$Internal.showAnotherParam()Ljava/lang/String;
  $sess.cmd2$.<init>(cmd2.sc:1)
  $sess.cmd2$.<clinit>(cmd2.sc:-1)

@ Bye!
```

As far as I can tell, it is fixed in 2.12.2, in both my minimization and your fuller test case. Maybe it's time for you to upgrade your Scala version 😛

**Comment by Alex-At-Home** [comment]
Hey @lihaoyi thanks for looking into this - fair enough not worrying about it since it's fixed in 2.12 (I figured it was because the underlying bug https://issues.scala-lang.org/browse/SI-9689 is fixed there). 

Should any thoughts as to how `scalacOptions ++= YnoLoadImplClass` in `sbt console` differs from `interp.configureCompiler(....)` in `ammonite` spring to mind, do feel free to ping them here :) I'm still stuck in 2.11 for a bit so if s an opportunity to workaround/fix in 2.11 occurs in the next few months I'll probably spend a bit of time on it.
